File: BETELITE_v3/BETELITE_v3/backend/routes/bets.js
// routes/bets.js
const express = require('express');
const { v4: uuid } = require('uuid');
const { db } = require('../services/firebase');
const bets = new Map();

module.exports = (io, engine) => {
  const r = express.Router();

  r.post('/place', async (req, res) => {
    const { userId, matchId, pick, odds, amount, currency } = req.body;
    if (!userId || !matchId || !pick || !odds || !amount)
      return res.status(400).json({ error: 'Missing fields' });

    const betAmount = parseFloat(amount);
    if (betAmount <= 0) return res.status(400).json({ error: 'Invalid amount' });

    // Validate and deduct balance atomically
    if (db) {
      const balRef = db.ref(`users/${userId}/wallet/balance`);
      const result = await balRef.transaction((currentBalance) => {
        const bal = currentBalance || 0;
        if (bal < betAmount) return; // Abort if insufficient
        return bal - betAmount;
      });
      if (!result.committed) {
        return res.status(400).json({ error: 'Insufficient balance' });
      }
    }

    const bet = {
      id: uuid(), userId, matchId, pick,
      odds: parseFloat(odds), amount: betAmount,
      currency: currency || 'NGN', status: 'live',
      potentialWin: Math.round(betAmount * parseFloat(odds)),
      placedAt: Date.now()
    };
    bets.set(bet.id, bet);

    // Persist bet to Firebase for settlement
    if (db) {
      db.ref(`bets/${bet.id}`).set(bet).catch(e => console.error('[BET] Persist error:', e.message));
    }

    res.json({ ok: true, bet });
  });

  r.get('/user/:userId', (req, res) => {
    const userBets = [...bets.values()].filter(b => b.userId === req.params.userId);
    res.json({ bets: userBets });
  });

  return r;
};

File: BETELITE_v3/BETELITE_v3/go_backend/routes/bets.go
package routes

import (
	"context"
	"fmt"
	"time"

	"github.com/gofiber/fiber/v2"

	"betelite-go/db"
	"betelite-go/middleware"
	"betelite-go/services"
)

func SetupBetRoutes(api fiber.Router) {
	bets := api.Group("/bets", middleware.AuthRequired())

	bets.Post("/place", func(c *fiber.Ctx) error {
		var req struct {
			MatchID string  `json:"matchId"`
			Pick    string  `json:"pick"` // home, away, draw
			Odds    float32 `json:"odds"`
			Amount  int64   `json:"amount"` // Note: kobo/pesewas
		}
		if err := c.BodyParser(&req); err != nil {
			return c.Status(400).JSON(fiber.Map{"error": "Invalid payload"})
		}

		if req.Amount <= 0 {
			return c.Status(400).JSON(fiber.Map{"error": "Amount must be greater than 0"})
		}

		// Verify match is live and taking bets
		match := services.Engine.GetMatch(req.MatchID)
		if match == nil || match.Status != "live" {
			return c.Status(400).JSON(fiber.Map{"error": "Match not available for betting"})
		}
		if match.Minute >= 80 {
			return c.Status(400).JSON(fiber.Map{"error": "Betting is closed for this match (80+ mins)"})
		}

		uid := middleware.GetUID(c)
		ctx := context.Background()

		// DB Transaction
		tx, err := db.Pool.Begin(ctx)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Database error"})
		}
		defer tx.Rollback(ctx)

		// Check balance
		var balance int64
		err = tx.QueryRow(ctx, "SELECT balance FROM users WHERE id = $1 FOR UPDATE", uid).Scan(&balance)
		if err != nil || balance < req.Amount {
			return c.Status(400).JSON(fiber.Map{"error": "Insufficient funds"})
		}

		// Deduct balance
		_, err = tx.Exec(ctx, "UPDATE users SET balance = balance - $1 WHERE id = $2", req.Amount, uid)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to update balance"})
		}

		// Insert Bet
		betID := fmt.Sprintf("bet_%d", time.Now().UnixNano())
		potentialWin := int64(float32(req.Amount) * req.Odds)

		_, err = tx.Exec(ctx, `INSERT INTO bets (id, user_id, match_id, pick, odds, amount, potential_win, status) 
			VALUES ($1, $2, $3, $4, $5, $6, $7, 'live')`,
			betID, uid, req.MatchID, req.Pick, req.Odds, req.Amount, potentialWin)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to place bet"})
		}

		// Log transaction
		_, err = tx.Exec(ctx, "INSERT INTO transactions (user_id, type, amount, ref_id) VALUES ($1, 'bet_place', $2, $3)",
			uid, -req.Amount, betID)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to log transaction"})
		}

		if err := tx.Commit(ctx); err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Transaction commit failed"})
		}

		return c.JSON(fiber.Map{"success": true, "betId": betID})
	})
}
